### PR TITLE
(feature) added date_passed & document_types to Legislations

### PR DESCRIPTION
**Pull request by kowal**
- Added `date_passed` and `document_types` (as tags) to `Legislations`.
- extracted some import helpers

you can re-import records via `bundle exec rake import:legislation`

Zrzut ekranu 2019-07-28 o 21 39 59

### CODE REVIEW

File: app/services/import_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module ImportHelpers
+  def find_location(country_iso)
+    Location.find_by!(iso: country_iso)
+  rescue StandardError
+    puts "Couldn't find Location with ISO: #{country_iso}"
+  end
+
+  def try_to_parse_date(date, expected_date_formats)
```
Comment (kowal) [inline app/services/import_helpers.rb]: maybe rename it to `find_date_with_format`?

File: app/models/concerns/taggable.rb [hunk 2]
```diff
@@ -1,3 +1,25 @@
+# To tag your model with let's say 'keywords', you need to
```
Comment (tsubik) [inline app/models/concerns/taggable.rb]: Thanks for this :+1:

File: app/models/document_type.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#
+#  id         :bigint           not null, primary key
+#  name       :string           not null
+#  type       :string           not null
+#  created_at :datetime         not null
+#  updated_at :datetime         not null
+#
+
+class DocumentType < Tag
+  DOCUMENT_TYPES = [
```
Comment (tsubik) [inline app/models/document_type.rb]: Hmm, not sure if we want to restrict that list

File: app/services/import_helpers.rb [hunk 4]
```diff
@@ -0,0 +1,17 @@
+module ImportHelpers
```
Comment (tsubik) [inline app/services/import_helpers.rb]: Good idea

File: app/services/import_helpers.rb (hunk 1 above)
Comment (tsubik) [reply app/services/import_helpers.rb]: You can change it, for me both are good.

File: app/models/document_type.rb (hunk 3 above)
Comment (tsubik) [reply app/models/document_type.rb]: Also, that list is funny now, I wouldn't pay too much attention to what is in the data currently. See there is document type `All`, they should clean this up, what they wanted to just tag Law with different types. Maybe we will have to extract some of those tags to be something else in the future (for example if application logic will depend on it, I suspect "Out of date" would be the first candidate)

File: app/services/import_helpers.rb (hunk 1 above)
Comment (kowal) [reply app/services/import_helpers.rb]: I will leave it as it is.

File: app/models/document_type.rb (hunk 3 above)
Comment (kowal) [reply app/models/document_type.rb]: yeah I agree, this list will definitely change. For now I can remove the validation, but eventually, I think we will need some constraint on "document type". I guess this is a more precise term than, let's say "category".

File: spec/controllers/admin/legislations_controller_spec.rb [hunk 5]
```diff
@@ -2,7 +2,16 @@
 
 RSpec.describe Admin::LegislationsController, type: :controller do
   let(:admin) { create(:admin_user) }
+  let(:location) { create(:location) }
+
   let!(:legislation) { create(:legislation) }
+  let(:valid_attributes) {
+    attributes_for(:legislation).merge(
```
Comment (kowal) [inline spec/controllers/admin/legislations_controller_spec.rb]: I couldn't find a way to include location by default in what's returned by `attributes_for` :/ then I found your approach from other models and used it here as well.

**Comment by kowal** [comment]
@tsubik more updates:
- `date_passed`: added filter + default sort order + fix parsing some dates in importer
- better CSV output
- added more specs

File: app/models/document_type.rb (hunk 3 above)
Comment (tsubik) [reply app/models/document_type.rb]: Not sure, I think right now those are like categories, and definitely, that list needs cleanup from their side. I would just remove validation for now.

File: app/services/import_helpers.rb (hunk 4 above)
Comment (tsubik) [inline app/services/import_helpers.rb]: We can move it to `app/services/concerns`. I'm not sure if that would require to add new `concern` directory to autoload_paths in `config/application.rb`, like `config.autoload_paths += Dir[Rails.root.join("app", "services", "concerns")]`

File: spec/controllers/admin/legislations_controller_spec.rb (hunk 5 above)
Comment (tsubik) [reply spec/controllers/admin/legislations_controller_spec.rb]: Yes, attributes_for are not working with associations, probably they have a good reason for that.

File: app/services/import_helpers.rb (hunk 4 above)
Comment (kowal) [reply app/services/import_helpers.rb]: I think concerns are meant to encapsulate a piece of domain logic, here we have 2 simple helper methods related to different things.

Maybe I add 2 different modules, which could be used like this:

```ruby
def normalize_date
  Import::Date.try_to_parse(date)
end

def attributes
  { location: Import::Location.find(..) }
end

```

Comment (kowal) [reply app/services/import_helpers.rb]: Also, I'd move all import classes to `services/import` as well, so we'll end up with `Import::Companies.new.call` and so on.

Comment (tsubik) [reply app/services/import_helpers.rb]: Ok.

Comment (kowal) [reply app/services/import_helpers.rb]: done, please re-review

**Pull request merged**